// @sapphire/framework/dist/arguments/CoreEnum.js
'use strict';

var pieces = require('@sapphire/pieces');
var enum_js = require('../lib/resolvers/enum.js');
var Argument_js = require('../lib/structures/Argument.js');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var _CoreArgument = class _CoreArgument extends Argument_js.Argument {
  constructor(context) {
    super(context, { name: "enum" });
  }
  run(parameter, context) {
    const resolved = enum_js.resolveEnum(parameter, { enum: context.enum, caseInsensitive: context.caseInsensitive });
    return resolved.mapErrInto(
      (identifier) => this.error({
        parameter,
        identifier,
        message: `The argument must have one of the following values: ${context.enum?.join(", ")}`,
        context
      })
    );
  }
};
__name(_CoreArgument, "CoreArgument");
var CoreArgument = _CoreArgument;
void pieces.container.stores.loadPiece({
  name: "enum",
  piece: CoreArgument,
  store: "arguments"
});

exports.CoreArgument = CoreArgument;
