// @sapphire/framework/dist/arguments/CoreFloat.js
'use strict';

var pieces = require('@sapphire/pieces');
var Identifiers_js = require('../lib/errors/Identifiers.js');
var float_js = require('../lib/resolvers/float.js');
var Argument_js = require('../lib/structures/Argument.js');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var _CoreArgument = class _CoreArgument extends Argument_js.Argument {
  constructor(context) {
    super(context, { name: "float" });
    this.messages = {
      [Identifiers_js.Identifiers.ArgumentFloatTooSmall]: ({ minimum }) => `The given number must be greater than ${minimum}.`,
      [Identifiers_js.Identifiers.ArgumentFloatTooLarge]: ({ maximum }) => `The given number must be less than ${maximum}.`,
      [Identifiers_js.Identifiers.ArgumentFloatError]: () => "The argument did not resolve to a valid decimal."
    };
  }
  run(parameter, context) {
    const resolved = float_js.resolveFloat(parameter, { minimum: context.minimum, maximum: context.maximum });
    return resolved.mapErrInto(
      (identifier) => this.error({
        parameter,
        identifier,
        message: this.messages[identifier](context),
        context
      })
    );
  }
};
__name(_CoreArgument, "CoreArgument");
var CoreArgument = _CoreArgument;
void pieces.container.stores.loadPiece({
  name: "float",
  piece: CoreArgument,
  store: "arguments"
});

exports.CoreArgument = CoreArgument;
