// @sapphire/framework/dist/arguments/CorePartialDMChannel.js
'use strict';

var pieces = require('@sapphire/pieces');
var partialDMChannel_js = require('../lib/resolvers/partialDMChannel.js');
var Argument_js = require('../lib/structures/Argument.js');

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var _CoreArgument = class _CoreArgument extends Argument_js.Argument {
  constructor(context) {
    super(context, { name: "partialDMChannel" });
  }
  run(parameter, context) {
    const resolved = partialDMChannel_js.resolvePartialDMChannel(parameter, context.message);
    return resolved.mapErrInto(
      (identifier) => this.error({
        parameter,
        identifier,
        message: "The argument did not resolve to a Partial DM channel.",
        context
      })
    );
  }
};
__name(_CoreArgument, "CoreArgument");
var CoreArgument = _CoreArgument;
void pieces.container.stores.loadPiece({
  name: "partialDMChannel",
  piece: CoreArgument,
  store: "arguments"
});

exports.CoreArgument = CoreArgument;
